identity-account/src/identity/identity_index.rs

```diff
@@ -79,6 +100,46 @@ impl Default for IdentityIndex {
   }
 }
 
+impl Clone for IdentityIndex {
+  fn clone(&self) -> Self {
+    let mut locks = HashMap::new();
+    self.data.values().for_each(|id| {
+      locks.insert(*id, Arc::new(RwLock::new(*id)));
+    });
+
+    Self {
+      data: self.data.clone(),
+      locks,
+    }
+  }
+}
+
+impl PartialEq for IdentityIndex {
+  fn eq(&self, other: &Self) -> bool {
+    self.data == other.data
+  }
+}
+
+impl<'de> serde::Deserialize<'de> for IdentityIndex {
+  fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+  where
+    D: serde::Deserializer<'de>,
+  {
+    Result::map(
+      serde::Deserialize::deserialize(deserializer),
+      |data: HashMap<IdentityTag, IdentityId>| {
+        let mut locks = HashMap::new();
+
+        for value in data.values() {
+          locks.insert(*value, Arc::new(RwLock::new(*value)));
+        }
+
+        IdentityIndex { data, locks }
+      },
+    )
+  }
+}
```

Comment:
Can we change `get_lock` to lazily create locks as-needed instead of creating them all at once here, or on `insert`? That would be a performance benefit too since one would not need locking at all if not using `update_identity` / `publish_changes`, and then only for those identities being updated.

I am a bit concerned with how the `IdentityIndex` is used by the `Storage` where it is cloned (in `MemStore`) or serialized then de-serialized (`Stronghold`) that one could circumvent the locks by using different copies of them. It may be better to store the lock map in the `Account` struct proper to avoid these issues altogether?

Reply:
You're right, this is better. I was previously worried about contention on the `RwLock<IdentityIndex>` in the account, since `get_lock` would then modify the index and require the lock to be acquired as a write lock rather than just a read one. But most likely the performance is better this way then adding locks for a large number of identities.

Yes theoretically possible, but prevent by the design of the account. The index is only cloned when returned by storage, and added when the account is created, and never modified as a whole. So for any given account no cloning happens, and thus this should be fine.

Having the locks in the account requires more locks, as you would have `RwLock<IdentityIndex>` and also a separate `RwLock<HashMap<IdentityId, Arc<RwLock<IdentityId>>>>`, which I wanted to avoid. One possibility is to remove the custom `Clone` impl and add a derive, in which case the `Arc` around the lock would be cloned, and thus your concern of getting around the locks when cloning would be alleviated.

Reply:
I think using something like `Dashmap<IdentityId, Arc<RwLock<IdentityId>>>` in the account wouldn't be too terrible.

`Clone` is only used by `MemStore`, the index is (de)serialized by the Stronghold storage so it would still suffer from that. I think in general this may not be a problem since, as you indicated, the `Account` itself is not cloned. I also do not see any cases where an `Account` re-loads the index from storage after the initial configuration but may be incorrect.

Overall the locking seems fine. I'll take another look as there may be places where a read lock is appropriate.